Masked channels

davidwalter2 opened the pull request:
- Add masked channels similar as it was done in combinetf1. 
- Masked channels are always added trailing while not masked channels are added leading
- To avoid calculating the yields in the masked channel during likelihood minimization the masked channels can be clipped in ```_compute_yields```
- Bin by bin uncertainties in the masked channel are treated in the same way with the exception that they are never profiled and don't contribute to the likelihood. 
- Previously the fitter had an attribute 'profile' as global variable to steer if the systematics are profiled or not. However, this is captured in the tf graph and would not be able to change one the graph is traced. But since his is expected to change between prefit and postfit, it was changed to use function parameter instead.

bendavid commented:
Please add something about the change of the "profile" behaviour to the PR description.

davidwalter2 commented:
> Please add something about the change of the "profile" behaviour to the PR description.

Done

bendavid commented on combinetf2/inputdata.py:
```diff
@@ -80,7 +103,18 @@ def __init__(self, filename, pseudodata=None, normalize=False):
                         ]
-                    }
+                    },
+                    "ch1": {
```
Should this channel be added in case there are no masked channels?

bendavid replied:
Also probably "masked" should appear in the channel name.

bendavid commented on combinetf2/tensorwriter.py:
```diff
@@ -96,8 +96,8 @@ def add_pseudodata(self, h, name=None, channel="ch0"):
 
-    def add_process(self, h, name, channel="ch0", signal=False):
-        self._check_hist_and_channel(h, channel)
+    def add_process(self, h, name, channel="ch0", signal=False, masked=False):
```
General question:  Why not just require the user to always call add_channel before calling add_process? (in this case the "masked" parameter would not be needed here, where it could be confusing/inconsistent)

davidwalter2 replied:
Yes we could be more restrictive and require the user to always call add_channel first. Probably less error prone.

bendavid commented:
One general comment which can be addressed in the future is that it could make sense to provide observed data for masked channels in some cases (but it would have to be optional clearly)

davidwalter2 commented on combinetf2/h5pyutils.py:
```diff
@@ -82,3 +82,27 @@ def writeSparse(indices, values, dense_shape, h5group, outname, maxChunkBytes=10
     return nbytes
+
+
+def simple_sparse_slice0end(in_sparse, end):
```
I tested and using
```
snormnorm_sparse = tf.sparse.slice(snormnorm_sparse, [0,0], [self.indata.nbins, self.indata.nproc])
```
Gives the same result. 
But running the full fit I get the error 'LookupError: gradient registry has no entry for: SparseSliceGrad' so we probably need to keep this helper function. I move it in a separate file "tfhelpers.py" if there is no objection.

davidwalter2 commented:
> One general comment which can be addressed in the future is that it could make sense to provide observed data for masked channels in some cases (but it would have to be optional clearly)

OK, I addressed all the points above and added an issue for this one: https://github.com/WMass/combinetf2/issues/18
Depending on how fast you review this PR and on my availability, I can either implement it on top of this PR or make a new one.
